File: programmers/java/src/level3/kakao/kakao2020/보석쇼핑.java
package level3.kakao.kakao2020;

import java.util.*;

public class 보석쇼핑 {
    public static void main(String[] args) {

    }
    public int[] solution(String[] gems) {
        Map<String, Integer> info = new HashMap<>();
        int left = 0;
        int right = 0;
        Set<String> gemsSet = new HashSet<>(Arrays.asList(gems));
        int gemsKinds = gemsSet.size();
        int[] answer = null;
        int cnt = 100005;

        while(left <= (gems.length - gemsKinds)) {
            int kind  = info.size();
            if(kind == gemsKinds) {
                if((right-left+1) < cnt) {
                    answer = new int[]{left+1, right};
                    cnt = right - left + 1;
                }
                info.compute(gems[left], (k, v) -> v -= 1);
                if(info.get(gems[left]) == 0) {
                    info.remove(gems[left]);
                }
                left++;
                continue;
            }
            if(right == gems.length) {
                break;
            }
            info.compute(gems[right], (k, v) -> (v==null) ? 1 : v+1);
            right++;
        }


        return answer;
    }
}

File: programmers/python/level3/kakao/2020/intern/보석쇼핑.py
# https://school.programmers.co.kr/learn/courses/30/lessons/67258?language=python3
from collections import defaultdict

def solution(gems):
    left = 0
    right = 0
    info = defaultdict(int)
    gems_kind = list(set(gems))
    gems_cnt = len(gems_kind)
    
    answer=[]
    cnt = 100005
    
    while left <= (len(gems) - gems_cnt) and right <= len(gems):     
        kind = len(info)
        if kind == gems_cnt:
            if (right - left + 1) < cnt:
                answer = [left+1, right]
                cnt = right - left + 1
            info[gems[left]] -= 1
            if info[gems[left]] == 0:
                del info[gems[left]]
            left += 1 
            continue
        
        if right == len(gems):
            break
        
        info[gems[right]] += 1
        right += 1 
    
    return answer
